domain/personallearningplan/src/main/kotlin/uk/gov/justice/digital/hmpps/domain/personallearningplan/service/GoalService.kt

```diff
@@ -84,6 +100,73 @@ class GoalService(
       }
   }
 
+  /**
+   * Archives a [Goal], identified by its `prisonNumber` and `goalReference`, from the specified [ArchiveGoalDto].
+   */
+  fun archiveGoal(prisonNumber: String, archiveGoalDto: ArchiveGoalDto): Either<ArchiveGoalProblem, Goal> {
```

Comment:
Happy to change this from Either/Problem style to some other but just wanted to demo it.

Reply:
It's definitely good to see this as a demo 👍 , and for the most part I can understand what it is doing, but I'll check out the branch in my IDE to have a proper play with it

I'm struggling with a couple of bits in this class .... the method chaining from `toOption` to `toEither` then a series of `flatMap` operations ...
It seems weird that you have to construct the ~exception~ problem class twice . eg. once on line 111 and again on line 115
And the calls to `right` and `left` (lines 115 and 118) are hard to read/follow

And the call to `Unit.right()` on line 147?

Reply:
Yes, appreciate it might be a bit jarring if you are not used to an either monad.

Reply:
pushed some updates with return types and using the constructors rather than short left/right methods.